### PR TITLE
[MLAS AArch64] SQNBitGemm optimization

### PR DESCRIPTION
**Pull request by edgchen1**
### Description

1. Add support for packing 4-bit values 32 at a time for CompInt8. 32 4-bit values can fit into a single 128-bit NEON register. For CompInt8, this enables a more efficient path for block sizes greater than or equal to 32. CompFp32 seems to do better with handling 16 elements at a time, so this 32-value packing is not used there.
    - Pack differently based on compute type. Adjust APIs to handle this.

2. Introduce template argument for whether to handle zero-point. This results in less code for the no zero-point (symmetric) case. However, there is a binary size increase due to the additional template instantiations.

### Motivation and Context

Continuing performance optimization.

### Benchmark Results

Baseline (d7ff81dfb7):

name|median_real_time|median_cpu_time
-|-|-
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|654354 ns|653462 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|613964 ns|613547 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|572359 ns|573044 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|620006 ns|610568 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|592421 ns|594580 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|569464 ns|569332 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|303690 ns|306242 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|242383 ns|242876 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|206066 ns|205532 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|284828 ns|283176 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|235312 ns|233447 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|205712 ns|205833 ns

Updated (1a0766471e):

name|median_real_time|median_cpu_time
-|-|-
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|638494 ns|633831 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|596357 ns|595869 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:1/real_time|561178 ns|551766 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|586340 ns|582134 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|576410 ns|573980 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:1/real_time|576217 ns|569123 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|293955 ns|296636 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|200667 ns|201180 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:0/ComputeType:4/real_time|136009 ns|133801 ns
SQNBITGEMM<4>/BlkLen:16/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|272861 ns|275627 ns
SQNBITGEMM<4>/BlkLen:32/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|190781 ns|188563 ns
SQNBITGEMM<4>/BlkLen:256/M:1/N:4096/K:4096/Threads:8/Symmetric:1/ComputeType:4/real_time|137862 ns|139672 ns

### CODE REVIEW

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 1]
```diff
@@ -15,14 +15,115 @@
...
+    return PackedQuantBDataSize;
+}
+
+void
+SQ4BitGemmPackQuantBData(
+    size_t N,
+    size_t K,
+    size_t BlkLen,
+    MLAS_SQNBIT_GEMM_COMPUTE_TYPE ComputeType,
+    const std::byte* QuantBDataBegin,
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:56]: >const std::byte [](http://example.com/codeflow?start=4&length=15)

nit: why not use a uint8_t?

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp [hunk 2]
```diff
@@ -15,14 +15,115 @@ Module Name:
...
+}
+
+}  // namespace
+
+//
+// General helpers.
+//
+
 namespace
 {
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:129]: std::index_sequence<Indices...> is not used. May remove it from parameter

**Review by yufenglee — DISMISSED** [review]
:shipit:

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:56]: I used std::byte here to clarify the intent.
uint8_t is 8-bit integer which allows operations like addition. std::byte is just a byte and we should treat the quantized B data as raw bytes.

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp:129]: it is used, though perhaps indirectly. it enables the deduction of the `size_t... Indices` template argument from the function parameter type. in `UnrolledLoop`, `std::make_index_sequence` is used to convert from `N` to `std::index_sequence<0, ..., N-1>`, which causes the value of `Indices` in `UnrolledLoopIterations` to be deduced as `0, ..., N-1`.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**